File: apps/sdkwork-im-h5/packages/sdkwork-im-h5-chat/src/services/chatMediaUploadService.ts
import type {
  DriveUploaderClient,
  DriveUploaderRequest,
  DriveUploaderUploadResult,
} from "@sdkwork/drive-app-sdk";
import { getDriveAppSdkClientWithSession } from "@sdkwork/im-h5-core/sdk";
import { readImH5IamSessionTokens } from "@sdkwork/im-h5-core/session";

export type ChatMediaUploadType = "image" | "audio" | "video" | "attachment";

export interface ChatDriveReference {
  driveUri: string;
  spaceId: string;
  nodeId: string;
}

const CHAT_DRIVE_APP_RESOURCE_TYPE = "im_conversation";
const CHAT_DRIVE_SCENE = "im";
const CHAT_DRIVE_SOURCE = "chat_message";

function resolveUploadUserId(): string {
  const session = readImH5IamSessionTokens();
  const userId =
    session?.context?.userId?.trim()
    || session?.user?.userId?.trim()
    || session?.user?.id?.trim();
  if (!userId) {
    throw new Error("Chat media upload requires user_id in the authenticated session.");
  }
  return userId;
}

function normalizeDriveUploadResult(result: DriveUploaderUploadResult): ChatDriveReference {
  const spaceId = result.uploadItem.spaceId || result.uploadSession.spaceId;
  const nodeId = result.uploadItem.nodeId || result.uploadSession.nodeId;
  if (!spaceId || !nodeId) {
    throw new Error("Drive uploader result is missing spaceId or nodeId.");
  }
  return {
    driveUri: `drive://spaces/${spaceId}/nodes/${nodeId}`,
    spaceId,
    nodeId,
  };
}

function getDriveUploader(): Pick<
  DriveUploaderClient,
  "uploadAudio" | "uploadAttachment" | "uploadImage" | "uploadVideo"
> {
  return getDriveAppSdkClientWithSession().uploader;
}

export async function uploadChatMediaFile({
  conversationId,
  file,
  type,
  originalFileName,
  contentType,
}: {
  conversationId: string;
  file: Blob;
  type: ChatMediaUploadType;
  originalFileName?: string;
  contentType?: string;
}): Promise<{ drive: ChatDriveReference; uploadResult: DriveUploaderUploadResult }> {
  const uploadRequest: DriveUploaderRequest = {
    file,
    userId: resolveUploadUserId(),
    appResourceType: CHAT_DRIVE_APP_RESOURCE_TYPE,
    appResourceId: conversationId,
    scene: CHAT_DRIVE_SCENE,
    source: CHAT_DRIVE_SOURCE,
    ...(originalFileName ? { originalFileName } : {}),
    ...(contentType ? { contentType } : {}),
  };

  const uploader = getDriveUploader();
  const uploadResult =
    type === "image"
      ? await uploader.uploadImage(uploadRequest)
      : type === "audio"
        ? await uploader.uploadAudio(uploadRequest)
        : type === "video"
          ? await uploader.uploadVideo(uploadRequest)
          : await uploader.uploadAttachment(uploadRequest);

  return {
    drive: normalizeDriveUploadResult(uploadResult),
    uploadResult,
  };
}

File: apps/sdkwork-im-flutter-mobile/packages/sdkwork_im_flutter_mobile_chat/lib/src/services/chat_media_upload_service.dart
import 'dart:typed_data';

import 'package:sdkwork_im_flutter_mobile_core/sdkwork_im_flutter_mobile_core.dart';

const _chatDriveAppResourceType = 'im_conversation';
const _chatDriveScene = 'im';
const _chatDriveSource = 'chat_message';

Future<DriveUploadReference> uploadChatMediaBytes({
  required String applicationPublicHttpUrl,
  required String conversationId,
  required Uint8List bytes,
  required String userId,
  String? accessToken,
  String? authToken,
  String type = 'image',
  String? originalFileName,
  String? contentType,
}) async {
  final client = DriveAppSdkClient.create(
    applicationPublicHttpUrl: applicationPublicHttpUrl,
    accessToken: accessToken,
    authToken: authToken,
  );
  final request = DriveUploaderRequest(
    bytes: bytes,
    userId: userId,
    appResourceType: _chatDriveAppResourceType,
    appResourceId: conversationId,
    scene: _chatDriveScene,
    source: _chatDriveSource,
    originalFileName: originalFileName,
    contentType: contentType,
    uploadProfileCode: type == 'image' ? 'image' : 'attachment',
  );
  return type == 'image'
      ? client.uploadImage(request)
      : client.uploadAttachment(request);
}

String resolveDriveApplicationBaseUrl(String configuredBaseUrl) {
  return resolveImApplicationBaseUrl(configuredBaseUrl);
}
